### PR TITLE
contrib: add helper to graph current configuration

### PR DESCRIPTION
**Pull request by liambeguin**
## Description

Add a sub-command to sync-places to generate a graphviz graph of the resources and places in a given environment.

A visual representation can help make sense of a large amount of information quickly.
This tried to put all relevant information in one place without being overly complicated.

I also tired to use a color blind friendly palette, not sure how that turned out.. 

I'm very open to modifying this based on feedback and recommendations.

I also tried to put this into a separate tool using fastapi to have the graph rendered in a browser directly, but ran into trouble. If that approach is preferred I can also push that instead, and maybe with a bit of support we can get it to work consistently.

## Testing
```console
$ ./labgrid/contrib/sync-places.py graph | dot -Tsvg -o labgrid-graph.svg
```
By default the command will just output a dot graph to stdout, piping that to `dot` we can generate images.

As noted in the docstring, this will generate a graph with:
- all resources, grouped by groupname and exporter.
- all places, with a list of tags (a darker shade of blue, if the place is acquired)
- solid edges between places and acquired resources
- dotted edges between places and unacquired resources
- edges between resources and places carry the match name if any.

here's a simple example of what the output looks like:
image

## Checklist
- [x] Documentation for the feature (documented in the docstring)
- [x] Tests for the feature 
- [x] PR has been tested

### CODE REVIEW

**Comment by Bastian-Krause** [comment]
Does this actually use anything from the `sync-places.py`? Maybe it would be better to have this as a separate script in `crontrib/` ?

**Comment by liambeguin** [comment]
> Does this actually use anything from `sync-places.py`?

Not much apart from the "setup". I thought it kind of made sense to group, but I'm happy to separate it. I'll update the PR

**Comment by liambeguin** [comment]
I ended up finding a solution to share the event loop between fastapi and labgrid so I also updated the script to serve the graph in a browser.

File: contrib/sync-places.py [hunk 1]
```diff
@@ -109,6 +109,13 @@ async def do_sync(session, args):
                     changed = True
 
             tags = config["places"][name].get("tags", {}).copy()
+            for k, v in tags.items():
+                if not isinstance(k, str):
+                    tags[str(k)] = v
+                    del(tags[k])
+                if not isinstance(v, str):
+                    tags[k] = str(v)
+
```
Comment (Bastian-Krause) [inline contrib/sync-places.py:118]: Now that the script resides separated in graph-places.py, this should be a separate PR.

File: contrib/graph-places.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
+import argparse
```
Comment (Bastian-Krause) [inline contrib/graph-places.py:1]: Since this is meant to be run as a script, we should add a shebang and set the executable bit.

File: pyproject.toml [hunk 3]
```diff
@@ -79,6 +79,10 @@ deb = [
     # labgrid[snmp]
     "pysnmp-lextudio>=4.4.12",
 ]
+contrib = [
+    "fastapi>=0.104.1",
+    "uvicorn>=0.24.0",
+]
```
Comment (Bastian-Krause) [inline pyproject.toml:85]: I don't think we should add dependencies for contrib scripts in our pyproject.toml. Better document this, either in a module doc string in the script itself or maybe add a README in contrib?

Comment (liambeguin) [reply pyproject.toml:85]: Okay, I'll add a requirements file in contrib/ with a readme.

A colleague also mentioned that this could be extended to serve as a web interface for users that aren't too familiar with command line interfaces. I might rename the script so that other routes can be added later.

**Review by Bastian-Krause — APPROVED** [review]

File: contrib/labgrid-webapp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    '''Generate a graphviz graph of the current configuration.
+
+    Graph displays:
+    - all resources, grouped by groupname and exporter.
+    - all places, with a list of tags
+    - solid edges between places and acquired resources
+    - dotted edges between places and unacquired resources
+    - edges between resources and places carry the match name if any.
+    '''
+    def res_node_attr(name: str, resource: Resource) -> dict[str, str]:
```
Comment (PaulVittorino) [inline contrib/labgrid-webapp:29]: I got the following error when running the code:

>   File "./contrib/labgrid-webapp", line 180, in get_graph
>     svg = await do_graph(session)
>   File "./contrib/labgrid-webapp", line 35, in do_graph
>     def res_node_attr(name: str, resource: Resource) -> dict[str, str]:
> TypeError: 'type' object is not subscriptable


Adding single quotes around the type annotation resolved it.
`def res_node_attr(name: str, resource: Resource) -> 'dict[str, str]':`

File: contrib/labgrid-webapp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                    <td border="0" align="left">Resource</td>
+                </tr>
+                <tr>
+                    <td port="cls">{resource.cls}</td>
+                    <td port="name" bgcolor="white">{name}</td>
+                </tr>
+            </table>>''',
+        }
+
+    def place_node_attr(name: str, place: Place) -> dict[str, str]:
```
Comment (PaulVittorino) [inline contrib/labgrid-webapp:44]: Same issue as above

`def place_node_attr(name: str, place: Place) -> 'dict[str, str]':`

File: contrib/README.rst [hunk 6]
```diff
@@ -0,0 +1,38 @@
+labgrid-webapp
+==============
+
+labgrid-webapp implements a browser interface to access some of labgrid's
+information.
+
```
Comment (PaulVittorino) [inline contrib/README.rst:6]: Should the `graphviz` command line tool requirement be mentioned?

File: contrib/labgrid-webapp (hunk 4 above)
Comment (liambeguin) [reply contrib/labgrid-webapp:29]: This comes from using different versions of Python. I'll update to use `typing.Dict` instead

File: contrib/README.rst (hunk 6 above)
Comment (liambeguin) [reply contrib/README.rst:6]: the `graphviz` library is listed in the `requirements-webapp.txt` file, with some instructions on how to get things running. would you want to see an explicit mention of the library? I'm not sure it's really useful, but we could add it.

Comment (PaulVittorino) [reply contrib/README.rst:6]: Providing the requirements file made running the application so simple I had forgotten about the `graphviz` Python library. My thought was explicitly calling out the system dependency may help future users.

Comment (liambeguin) [reply contrib/README.rst:6]: I might be missing something, but I don't think there's a dependency on the `graphviz` system package. Did you have to install it to get things working? I might've had it installed already.

Comment (PaulVittorino) [reply contrib/README.rst:6]: https://graphviz.readthedocs.io/en/stable/manual.html does mention needing an installation of Graphviz. I got a very lengthy error from the `graphviz` Python library when I tried viewing the graph web page before I installed Graphviz.

Comment (liambeguin) [reply contrib/README.rst:6]: Oh! thanks for pointing that out. I added a note in the README.

**Comment by liambeguin** [comment]
I just rebased this MR on the latest master. If there are no other comments I think this would be ready to merge.

**Review by Bastian-Krause — APPROVED** [review]
